
option optimize_for = SPEED;
option java_package = "mooc";

// general status. The NO[word] values are 'failure for cause' and FAILURE is w/o cause.
enum PokeStatus {
   UKNOWN = 1;
   SUCCESS = 2;
   NOFOUND = 100;
   NOINCOMPLETE = 101;
   NOAUTH = 102;
   NOCONNECTION = 103;
   NOREACHABLE = 104;
   NORESOURCE = 105;
   FAILURE = 199;
}

// data - a hierartical container to send/receive name-value pairs
message NameValueSet {
    // a node should not have a value though this is not enforced
    required NodeType node_type = 1;

    optional string name = 2;
    optional string value = 3;

    // nested NVPs
    repeated NameValueSet node = 4;
    
    enum NodeType {
    	NODE = 1;
    	VALUE = 2;
    }
}

// data - a logical seperation of data/processing capabilities (same as a tentant)
message NameSpace {
   // internally created or supplied by caller when performing an action on a NS
   required int64 ns_id = 2 [default = -1]; 

   optional string name = 3;
   optional string desc = 4;
   optional int64 created = 5;
   optional int64 last_modified = 6;
   optional string owner = 7;

   // additional properties of the space
   optional NameValueSet properties = 8;
}

// data - a course description
message CourseDesc {
   required string name_space = 1;
   required int64 owner_id = 2;
   required string course_id = 3;
   required CourseCode status = 4;
   optional NameValueSet options = 5;

   enum CourseCode {
       COURSEUNKNOWN = 1;
       COURSERECEIVED = 2;
       COURSEQUEUED = 3;
       COURSERUNNING = 4;
       COURSEKILLED = 100;
       COURSEFAILED = 101;
   }
}

// request/response - check to see if a node is accepting requests
message Ping {
   required int32 number = 1;
   required string tag = 2;
}

// request - perform an operation to the namespace
message NameSpaceOperation {
   required SpaceAction action = 1;
   optional int64 ns_id = 2 [default = -1]; 
   optional NameSpace data = 3;

   // additional options for operation
   optional NameValueSet options = 4;

   enum SpaceAction {
      ADDSPACE = 1;
      UPDATESPACE = 2;
      REMOVESPACE = 3;
      LISTSPACES = 4;
   }
}

// response - reply from a space operation
message NameSpaceStatus {
   required PokeStatus status = 1;

   // data pertaining to the space operation
   repeated NameSpace data = 2;
}

// request - a course operation
message CourseOperation {
   required CourseAction action = 1;
  
   // ID for actions on known courses
   optional string course_id = 2;

   // new course to add
   optional CourseDesc data = 3;

   enum CourseAction {
      ADDCOURSE = 1;
      STOPCOURSE = 2;
      REMOVECOURSE = 3;
      LISTCOURSES = 4;
   }
}

// request - a node submits a course to be ran on the cluster
message CourseProposal {
   required string name_space = 1;
   required int64 owner_id = 2;
   required string course_id = 3;

   // weight is a scalar value representing the expected load for 
   // performing this course (e.g., 1 = light, 10 = intensive)
   required int32 weight  = 4;

   // properties of the course
   optional NameValueSet options = 5;
}

// request - a node's bid for running a course (a response in the form of a request)
message CourseBid {
   required string name_space = 1;
   required int64 owner_id = 2;
   required string course_id = 3;

   // a weighted bid -> this helps to remove ties in systems with a
   // even number of voters 
   required int32 bid = 4;
}


// response - the status of a course
message CourseStatus {
    required string course_id = 1;
    required PokeStatus status = 2;
    required CourseDesc.CourseCode course_state = 3;

    // data associated with a course request
    repeated CourseDesc data = 4;
}

message RoutingPath {
    required string node = 1;
    required int64 time = 2;
}

message Header {  
    required Routing routing_id = 1;
    required string originator = 2;
    optional string tag = 3;
    optional int64 time = 4;
    
    // while we can use extensions, it is not a clean inheritance 
    // implementation and since requests and responses differ only 
    // by these two fields, a new message is not created.

    optional PokeStatus reply_code = 5;
    optional string reply_msg = 6;
    
    // the path a message takes - to handle cycles in the network
    repeated RoutingPath path = 7;

    // if the message is for a specific node, this will be set
    optional string toNode = 8;

    // catch all for options
    repeated NameValueSet options = 9;

    enum Routing {
        PING = 2;
        NAMESPACES = 3;
        COURSES = 4;
        REPORTS = 10;
        MANAGE = 100;
        }
}

// these are commands from external clients
message Payload {
    // requests & replies
    optional Ping ping = 1;

    // requests only
    optional NameSpaceOperation space_op = 2;
    optional CourseOperation course_op = 3;

    // replies only
    optional CourseStatus course_status = 6;
    optional NameSpaceStatus space_status = 7;
   
    // TBD
    // reports
}

message Request {
    required Header header = 1;
    required Payload body = 2;
}

message Heartbeat {
    required string node_id = 1;
    required int64 time_ref = 2;
}

message Network {
    required string node_id = 1;
    required NetworkAction action = 2;

    enum NetworkAction {
       NODEJOIN = 1;
       NODELEAVE = 2;
       NODEDEAD = 3;
       CREATEMAP = 55;
       ANNOUNCE = 56;
       SHUTDOWN = 999;
   }
}

message LeaderElection {
   required string node_id = 1;
   required string ballot_id = 2;
   required string desc = 4;
   required VoteAction vote = 5;
   optional int64 expires = 6;

   enum VoteAction {
      ELECTION = 1;
      NOMINATE = 2;
      ABSTAIN = 3;
      DECLAREWINNER = 4;
      DECLAREVOID = 5;
      LEADER = 6;
   }
}

// internal communication
message Management {
   // a node joined or left the network
   optional Network graph = 1;

   // status updates between adjacent nodes
   optional Heartbeat beat = 2;

   // declare an election for a new leader 
   optional LeaderElection election = 3;

   // course management
   optional CourseProposal course_propose = 4;
   optional CourseBid course_bid = 5;
}	
